### PR TITLE
Fix nits in rpc error display.

### PR DESCRIPTION
**Pull request by marcio-diaz**
I took a quick look at https://github.com/paritytech/substrate/issues/6129 and couldn't reproduce. But there are some nits in the error display that may be related.

**Query 1**
```
{"id":1, "jsonrpc":"2.0", "method": "author_submitAndWatchExtrinsic", "params":["0x2d0284ff78b6dd81f9f55c08fdedb28e5e78e44a1ce6568164d4bd43fa4630a7a388592701e614839715c97beb90ed62b742f12fdcfd3f56d9ce72fefc32588799026c4e0154497063f375d5f49d5a03a921f67a74d0457c97f736b2a8c6ad2e7919e658850000000600ff8eaf04151687736326c9fea17e25fc5287613693c912909cb226aa4794f26a48227e"]}
```

Before changes:
```
< {"jsonrpc":"2.0","error":{"code":1002,"message":"Verification Error: Execution(ApiError(\"Could not convert parameter `tx` between node and runtime: Error decoding field Call :: Balances.0\"))","data":"RuntimeApi(\"Execution(ApiError(\\\"Could not convert parameter `tx` between node and runtime: Error decoding field Call :: Balances.0\\\"))\")"},"id":1}
```

After:
```
< {"jsonrpc":"2.0","error":{"code":1002,"message":"Verification Error: Execution: Could not convert parameter `tx` between node and runtime: Error decoding field Call :: Balances.0","data":"RuntimeApi(\"Execution: Could not convert parameter `tx` between node and runtime: Error decoding field Call :: Balances.0\")"},"id":1}
```
(actually display a readable `message`)

**Query 2:**
```
> {"id":1, "jsonrpc":"2.0", "method": "author_submitAndWatchExtrinsic", "params":["0x2d0284ff78b6dd81f9f55c08fdedb28e5e78e44a1ce6568164d4bd43fa4630a7a388592701e614839715c97beb90ed62b742f12fdcfd3f56d9ce72fefc32588799026c4e0154497063f375d5f49d5a03a921f67a74d0457c97f736b2a8c6ad2e7919e658850000000600ff8eaf04151687736326c9fea17e25fc5287613693c912909cb226aa4794f26a48214e"]}
```

Before changes:
```
< {"jsonrpc":"2.0","error":{"code":1010,"message":"Invalid Transaction","data":"BadProof"},"id":1}
```

After:
```
< {"jsonrpc":"2.0","error":{"code":1010,"message":"Invalid Transaction: Transaction has a bad signature","data":"BadProof"},"id":1}
```

(added the description of `InvalidTransaction` enum to `message`)

### CODE REVIEW

File: primitives/api/src/lib.rs [hunk 1]
```diff
@@ -344,7 +344,7 @@ pub trait ConstructRuntimeApi<Block: BlockT, C: CallApiAt<Block>> {
 #[cfg(feature = "std")]
 pub trait ApiErrorExt {
 	/// Error type used by the runtime apis.
-	type Error: std::fmt::Debug + From<String>;
+	type Error: std::fmt::Debug + std::fmt::Display + From<String>;
```
Comment (bkchr) [inline primitives/api/src/lib.rs:347]: Please revert this change.

File: primitives/api/src/lib.rs [hunk 2]
```diff
@@ -453,7 +453,7 @@ pub struct CallApiAtParams<'a, Block: BlockT, C, NC, Backend: StateBackend<HashF
 #[cfg(feature = "std")]
 pub trait CallApiAt<Block: BlockT> {
 	/// Error type used by the implementation.
-	type Error: std::fmt::Debug + From<String>;
+	type Error: std::fmt::Debug + std::fmt::Display + From<String>;
```
Comment (bkchr) [inline primitives/api/src/lib.rs:456]: This one as well.

File: client/transaction-pool/src/api.rs [hunk 3]
```diff
@@ -105,7 +105,7 @@ where
 				#[allow(deprecated)] // old validate_transaction
 				runtime_api.validate_transaction_before_version_2(&at, uxt)
 			};
-			let res = res.map_err(|e| Error::RuntimeApi(format!("{:?}", e)));
+			let res = res.map_err(|e| Error::RuntimeApi(format!("{}", e)));
```
Comment (bkchr) [inline client/transaction-pool/src/api.rs:108]: You need to change the where bound above. From:
```
sp_api::ApiErrorFor<Client, Block>: Send
```

to

```
sp_api::ApiErrorFor<Client, Block>: Send + std::fmt::Display
```

**Comment by marcio-diaz** [comment]
> And in general I find it weird that we now send the same message twice in the answer. Why can we not replace data with the display formatted error?



> And in general I find it weird that we now send the same message twice in the answer. Why can we not replace data with the display formatted error?

I didn't change it, it was like that from before. It locks aligned with the spec https://www.jsonrpc.org/specification#error_object. I can change it if you think is better, but what do we put in message?

**Comment by bkchr** [comment]
I mean instead of:
```
< {"jsonrpc":"2.0","error":{"code":1010,"message":"Invalid Transaction: Transaction has a bad signature","data":"BadProof"},"id":1}
```

we return:

```
< {"jsonrpc":"2.0","error":{"code":1010,"message":"Invalid Transaction","data":"Transaction has a bad signature"},"id":1}
```

**Comment by marcio-diaz** [comment]
> I mean instead of:
> 
> ```
> < {"jsonrpc":"2.0","error":{"code":1010,"message":"Invalid Transaction: Transaction has a bad signature","data":"BadProof"},"id":1}
> ```
> 
> we return:
> 
> ```
> < {"jsonrpc":"2.0","error":{"code":1010,"message":"Invalid Transaction","data":"Transaction has a bad signature"},"id":1}
> ```

I did it like that at first, but there

> I mean instead of:
> 
> ```
> < {"jsonrpc":"2.0","error":{"code":1010,"message":"Invalid Transaction: Transaction has a bad signature","data":"BadProof"},"id":1}
> ```
> 
> we return:
> 
> ```
> < {"jsonrpc":"2.0","error":{"code":1010,"message":"Invalid Transaction","data":"Transaction has a bad signature"},"id":1}
> ```

Oh, I see. i did it like that at first, but then I saw `InvalidTransaction::Custom(_) => "InvalidTransaction custom error"` and though we may be missing info. Also it keeps consistency with verification error.

File: client/transaction-pool/src/api.rs (hunk 3 above)
Comment (marcio-diaz) [reply client/transaction-pool/src/api.rs:108]: Thanks!

Comment (niklasad1) [reply client/transaction-pool/src/api.rs:108]:
```suggestion
			let res = res.map_err(|e| Error::RuntimeApi(e.to_string()));
```

**Comment by bkchr** [comment]
> i did it like that at first, but then I saw InvalidTransaction::Custom(_) => "InvalidTransaction custom error" and though we may be missing info. Also it keeps consistency with verification error.

Not sure what you mean by this.

**Comment by marcio-diaz** [comment]
> > i did it like that at first, but then I saw InvalidTransaction::Custom(_) => "InvalidTransaction custom error" and though we may be missing info. Also it keeps consistency with verification error.
> 
> Not sure what you mean by this.

I didn't want to lose whatever is inside `Custom(_)`.

**Comment by bkchr** [comment]
Than add it to the match arm?

**Comment by marcio-diaz** [comment]
> Than add it to the match arm?

How? Is there a quick way that justifes it?

**Comment by bkchr** [comment]
```
Error::Pool(PoolError::InvalidTransaction(InvalidTransaction::Custom(e))) => custom,
Error::Pool(PoolError::InvalidTransaction(e)) => default
```

**Comment by marcio-diaz** [comment]
> ```
> Error::Pool(PoolError::InvalidTransaction(InvalidTransaction::Custom(e))) => custom,
> Error::Pool(PoolError::InvalidTransaction(e)) => default
> ```

A bit ugly, but I'll change it :+1:

**Comment by bkchr** [comment]
Not sure why this is "ugly". That is how it works. You can also just write a proper `Display` implementation for `InvalidTransaction`.

**Comment by marcio-diaz** [comment]
Yes, a proper `Display` implementation would be better. I could give it a try in a follow up.

**Comment by marcio-diaz** [comment]
I pushed your suggestion Basti, let me know if it's ok now. Thanks!

File: client/rpc-api/src/author/error.rs [hunk 4]
```diff
@@ -114,10 +115,18 @@ impl From<Error> for rpc::Error {
 				message: format!("Verification Error: {}", e).into(),
 				data: Some(format!("{:?}", e).into()),
 			},
-			Error::Pool(PoolError::InvalidTransaction(e)) => rpc::Error {
+			Error::Pool(PoolError::InvalidTransaction(InvalidTransaction::Custom(e))) => rpc::Error {
 				code: rpc::ErrorCode::ServerError(POOL_INVALID_TX),
 				message: "Invalid Transaction".into(),
-				data: serde_json::to_value(e).ok(),
+				data: Some(e.into()),
```
Comment (bkchr) [inline client/rpc-api/src/author/error.rs:121]:
```suggestion
				data: Some(format!("Custom error: {}", e)),
```

**Comment by bkchr** [comment]
Ty
